perf/uncore_libpfm.go

```diff
@@ -159,41 +162,63 @@ func NewUncoreCollector(cgroupPath string, events PerfEvents, topology []info.No
 }
 
 func (c *uncoreCollector) setup(events PerfEvents, devicesPath string) error {
-	var err error
 	readUncorePMUs, err := getUncorePMUs(devicesPath)
 	if err != nil {
 		return err
 	}
 
-	// Maping from event name, pmu type, cpu.
-	c.cpuFiles = make(map[string]map[string]map[int]readerCloser)
+	c.cpuFiles = make(map[int]map[string]group)
 	c.events = events.Uncore.Events
 	c.eventToCustomEvent = parseUncoreEvents(events.Uncore)
 	c.cpuFilesLock.Lock()
 	defer c.cpuFilesLock.Unlock()
 
-	for _, group := range c.events {
-		if len(group) > 1 {
-			klog.Warning("grouping uncore perf events is not supported!")
-			continue
+	for i, group := range c.events {
+		// Check what PMUs are needed.
+		groupPMUs, err := parsePMUs(group, readUncorePMUs, c.eventToCustomEvent)
+		if err != nil {
+			return err
 		}
 
-		eventName, pmuPrefix := parseEventName(string(group[0]))
+		err = checkGroup(group, groupPMUs)
+		if err != nil {
+			return err
+		}
 
-		var err error
-		customEvent, ok := c.eventToCustomEvent[group[0]]
-		if ok {
-			if customEvent.Type != 0 {
-				pmus := obtainPMUs("uncore", readUncorePMUs)
-				err = c.setupRawNonGroupedUncore(customEvent, pmus)
+		// CPUs file descriptors of group leader needed for perf_event_open.
+		leaderFileDescriptors := make(map[string]map[uint32]int)
+		for _, pmu := range readUncorePMUs {
+			leaderFileDescriptors[pmu.name] = make(map[uint32]int)
+			for _, cpu := range pmu.cpus {
+				leaderFileDescriptors[pmu.name][cpu] = -1
+			}
+		}
```

Comment:
You initialize `leaderFileDescriptors`, why? Can't it be handled in a way similar to one that I suggested for `collector`?